### PR TITLE
pallet-staking: Add extrinsic `force_apply_min_commission`

### PR DESCRIPTION
**Pull request by emostov**
This PR adds the extrinsic `force_apply_min_commission`. The extrinsic can only be called by root and will increase any commission that is less than the `MinCommission`.

polkadot companion: https://github.com/paritytech/polkadot/pull/4881

This PR also improves the docs for `try_mutate_exists` to specify that the closure will only be called with `Some` if the value exists

### CODE REVIEW

File: frame/staking/src/pallet/mod.rs
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:1635]:
```suggestion
				prefs.commission = prefs.commission.max(min_commission);
```

**Review by kianenigma — APPROVED** [review]

File: frame/staking/src/pallet/mod.rs
Comment (ggwpez) [inline frame/staking/src/pallet/mod.rs:1628]: There is no guarantee that the caller passes the correct value here, right?  
But since it is only root callable, this is probably not a problem.

**Review by ggwpez — APPROVED** [review]
I like it! But please wait for another approval.

File: frame/staking/src/tests.rs
Comment (ggwpez) [inline frame/staking/src/tests.rs:4747]: When I add this to the end of the test, it fails. Why?  
```rust
assert_ok!(Staking::force_apply_min_commission(Origin::root(), 15));
assert_eq!(validators(), vec![(31, prefs(15)), (21, prefs(15)), (11, prefs(15))]);
```

**Comment by shawntabrizi** [comment]
Generally these kinds of unbounded loops are an anti-pattern for the runtime.

I get that it is a root function and all, but I don't see why this has to be this way.

You could just as easily make a function which updates just a single validator, and let it be called by anyone. Then someone can write a batch script, or ping old validators one by one to update them.

File: frame/staking/src/tests.rs
Comment (shawntabrizi) [reply frame/staking/src/tests.rs:4747]: the second param in `force_apply_min_comission` is not the commission amount. You will find min commission as a configuration of the pallet.

**Comment by kianenigma** [comment]
> Generally these kinds of unbounded loops are an anti-pattern for the runtime.
> 
> I get that it is a root function and all, but I don't see why this has to be this way.
> 
> You could just as easily make a function which updates just a single validator, and let it be called by anyone. Then someone can write a batch script, or ping old validators one by one to update them.

fwiw we generally assume that we can absolutely iterate over all validators all over the staking system, which is why I recommended this PR this way for the sake of simplicity. 

But I see your point: Let's not build more on top of a bad assumption. Batching a transaction that sets a single validators is also easy enough.

**Comment by emostov** [comment]
> > Generally these kinds of unbounded loops are an anti-pattern for the runtime.
> > I get that it is a root function and all, but I don't see why this has to be this way.
> > You could just as easily make a function which updates just a single validator, and let it be called by anyone. Then someone can write a batch script, or ping old validators one by one to update them.
> 
> fwiw we generally assume that we can absolutely iterate over all validators all over the staking system, which is why I recommended this PR this way for the sake of simplicity.
> 
> But I see your point: Let's not build more on top of a bad assumption. Batching a transaction that sets a single validators is also easy enough.

In the scenario the community wants to set the min commission and and force apply it in the same batch call, setting single validators is sub optimal. Once the call is crafted and proposed, validators who previously had higher commissions can reduce their commission to below the min, effectively circumventing the force apply. You could specify every validator knowing it would be a no-op if their commission was high enough, but I am not sure that approach would scale if the number of validators becomes very large.

My plan is to change this so the number of iterations is bounded by the weight input param.

File: frame/staking/src/pallet/mod.rs
Comment (emostov) [inline frame/staking/src/pallet/mod.rs:1638]: @shawntabrizi does this suffice for bounding? Here is my rational for not just force applying to single validators: https://github.com/paritytech/substrate/pull/10786#issuecomment-1029196447

Comment (shawntabrizi) [reply frame/staking/src/pallet/mod.rs:1638]: not a good solution.

We will only be able to iterate over the same first n people. How should we touch the last ones?

**Comment by kianenigma** [comment]
> effectively circumventing the force apply

But in this case they already increased their commission and the objective is achieved. 

I think a batch call is less of the evil and simple enough. 

It is not that we are going to do this every hour, so we don't need to care too much about it being the most efficient.

**Comment by emostov** [comment]
> But in this case they already increased their commission and the objective is achieved.

What I was trying to say is that they can lower their commission after the call has been put up as a referenda. However I do think the approach of specifying a single validator is more extensible in the long term.

I update the impl in [4077606](https://github.com/paritytech/substrate/pull/10786/commits/407760601d4692630e661a618813d43d6ab11c58) to only take 1 stash as an arg, so needs a re-review @kianenigma @shawntabrizi

File: frame/staking/src/tests.rs
Comment (kianenigma) [inline frame/staking/src/tests.rs:4724]: commented code.

**Review by kianenigma — APPROVED** [review]
still LGTM 👍

File: frame/staking/src/tests.rs
Comment (emostov) [reply frame/staking/src/tests.rs]: 🤦 fixed in [18c0089](https://github.com/paritytech/substrate/pull/10786/commits/18c00897d5e1aa17667f536f8fc5459a88699adf)

File: frame/staking/src/pallet/mod.rs
Comment (shawntabrizi) [inline frame/staking/src/pallet/mod.rs:1627]: should be `ensure_signed`?

File: frame/staking/src/pallet/mod.rs
Comment (shawntabrizi) [inline frame/staking/src/pallet/mod.rs:1631]: seems like this can be done in one storage call

Comment (emostov) [reply frame/staking/src/pallet/mod.rs:1631]: I'll double check, but as far as I can tell this is a `ValueQuery` and the `OnEmpty` impl for this returns the default validator prefs; so if we didn't check that key existed then we would create an entry here because we would get the default, modify it and then write it

Comment (emostov) [reply frame/staking/src/pallet/mod.rs:1631]: If the map used an `OptionQuery` we could def skip the `contains_key`

File: frame/staking/src/pallet/mod.rs
Comment (emostov) [reply frame/staking/src/pallet/mod.rs]: [5ec80f2](https://github.com/paritytech/substrate/pull/10786/commits/5ec80f20623f8c9507dfd455b0db77bda26e3109)

File: frame/staking/src/pallet/mod.rs
Comment (shawntabrizi) [reply frame/staking/src/pallet/mod.rs:1632]: There should be a way to backdoor this since in the DB, everything is literally an option and not exposing an option to the runtime developer is just an API thing

**Review by shawntabrizi — APPROVED** [review]

**Comment by emostov** [comment]
@shawntabrizi 

- doc wording update [0d90a92](https://github.com/paritytech/substrate/pull/10786/commits/0d90a92a052acbfe69fba0ffcbb0806aeb24e410) 
- extrinsic change to return dispatch error if stash doesn't exist https://github.com/paritytech/substrate/pull/10786/commits/0d90a92a052acbfe69fba0ffcbb0806aeb24e410

File: frame/staking/src/pallet/mod.rs
Comment (shawntabrizi) [inline frame/staking/src/pallet/mod.rs:669]:
```suggestion
```

There is already `Error::<T>::NotStash`

File: frame/staking/src/pallet/mod.rs
Comment (shawntabrizi) [inline frame/staking/src/pallet/mod.rs:1639]:
```suggestion
					.ok_or(Error::<T>::NotStash)
```

File: frame/staking/src/tests.rs
Comment (shawntabrizi) [inline frame/staking/src/tests.rs:4745]:
```suggestion
			Error::<Test>::NotStash
```

**Review by shawntabrizi — APPROVED** [review]

**Comment by emostov** [comment]
bot merge

**Pull request merged**
